File: Desafios AD-Hoc/Set 1090/Set.py
from itertools import combinations
from collections import Counter

def parse_card(card):
    number_map = {"um": 1, "dois": 2, "tres": 3}
    shape_map = {
        "circulo": "circulo", "circulos": "circulo",
        "quadrado": "quadrado", "quadrados": "quadrado",
        "triangulo": "triangulo", "triangulos": "triangulo"
    }
    number, shape = card.split()
    return (number_map[number], shape_map[shape])

def is_valid_set(cards):
    for i in range(2):  # Verificar número e figura
        values = {card[i] for card in cards}
        if len(values) == 2:  # Não pode ter exatamente dois valores iguais
            return False
    return True

def count_sets(cards):
    parsed_cards = [parse_card(card) for card in cards]
    card_counts = Counter(parsed_cards)
    count = 0
    for combo in combinations(parsed_cards, 3):
        if is_valid_set(combo):
            # Se todas as cartas forem iguais, só deve contar se houver pelo menos 3 iguais
            if len(set(combo)) == 1:
                if card_counts[combo[0]] >= 3:
                    count += 1
            else:
                count += 1
    return count

def main():
    while True:
        n = int(input())
        if n == 0:
            break
        cards = [input().strip() for _ in range(n)]
        print(count_sets(cards))

File: Desafios AD-Hoc/Set 1090/Set.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <set>
#include <algorithm>

using namespace std;

pair<int, string> parseCard(const string& card) {
    map<string, int> numberMap = {{"um", 1}, {"dois", 2}, {"tres", 3}};
    map<string, string> shapeMap = {
        {"circulo", "circulo"}, {"circulos", "circulo"},
        {"quadrado", "quadrado"}, {"quadrados", "quadrado"},
        {"triangulo", "triangulo"}, {"triangulos", "triangulo"}
    };
    size_t spacePos = card.find(' ');
    string number = card.substr(0, spacePos);
    string shape = card.substr(spacePos + 1);
    return {numberMap[number], shapeMap[shape]};
}

bool isValidSet(const vector<pair<int, string>>& combo) {
    set<int> numbers;
    set<string> shapes;
    for (const auto& card : combo) {
        numbers.insert(card.first);
        shapes.insert(card.second);
    }
    return (numbers.size() != 2 && shapes.size() != 2);
}

int countSets(const vector<pair<int, string>>& cards) {
    int count = 0;
    int n = cards.size();
    for (int i = 0; i < n; ++i) {
        for (int j = i + 1; j < n; ++j) {
            for (int k = j + 1; k < n; ++k) {
                vector<pair<int, string>> combo = {cards[i], cards[j], cards[k]};
                if (isValidSet(combo)) {
                    if (cards[i] == cards[j] && cards[j] == cards[k]) {
                        int occurrences = count_if(cards.begin(), cards.end(), [&](const pair<int, string>& c) {
                            return c == cards[i];
                        });
                        if (occurrences >= 3) {
                            count++;
                        }
                    } else {
                        count++;
                    }
                }
            }
        }
    }
    return count;
}

int main() {
    while (true) {
        int n;
        cin >> n;
        if (n == 0) break;
        cin.ignore();
        vector<pair<int, string>> cards;
        for (int i = 0; i < n; ++i) {
            string card;
            getline(cin, card);
            cards.push_back(parseCard(card));
        }
        cout << countSets(cards) << endl;
    }
    return 0;
}
